src/logging.c

```diff
@@ -24,81 +24,260 @@ extern "C"
+    &g_rcutils_logging_severities_map, name, name_length);
+  if (!severity_string) {
+    if (rcutils_string_map_key_existsn(&g_rcutils_logging_severities_map, name, name_length)) {
+      // The severity threshold has been specified but couldn't be retrieved.
+      return -1;
+    }
+    return RCUTILS_LOG_SEVERITY_UNSET;
+  }
+  int severity;
+  if (strcmp("DEBUG", severity_string) == 0) {
+    severity = RCUTILS_LOG_SEVERITY_DEBUG;
+  } else if (strcmp("INFO", severity_string) == 0) {
+    severity = RCUTILS_LOG_SEVERITY_INFO;
+  } else if (strcmp("WARN", severity_string) == 0) {
+    severity = RCUTILS_LOG_SEVERITY_WARN;
+  } else if (strcmp("ERROR", severity_string) == 0) {
+    severity = RCUTILS_LOG_SEVERITY_ERROR;
+  } else if (strcmp("FATAL", severity_string) == 0) {
+    severity = RCUTILS_LOG_SEVERITY_FATAL;
+  } else {
+    fprintf(
+      stderr,
+      "Logger has an invalid severity threshold: %s\n", severity_string);
+    severity = RCUTILS_LOG_SEVERITY_UNSET;
+  }
+  return severity;
+}
+
+int rcutils_logging_get_logger_effective_threshold(const char * name)
+{
+  RCUTILS_LOGGING_AUTOINIT
+  size_t substring_length = strlen(name);
+  while (true) {
+    int severity = rcutils_logging_get_logger_severity_thresholdn(name, substring_length);
+    if (-1 == severity) {
+      fprintf(
+        stderr,
+        "Error getting effective severity threshold of logger '%s'\n", name);
+      return -1;
+    }
+    if (severity != RCUTILS_LOG_SEVERITY_UNSET) {
+      return severity;
+    }
+    // Determine the next ancestor's FQN by removing the child's name.
+    size_t index_last_separator = rcutils_find_lastn(name, '.', substring_length);
+    if (index_last_separator == substring_length) {
+      // There are no more separators in the substring.
+      // The name we just checked was the last that we needed to, and it was unset.
+      break;
+    }
+    // Shorten the substring to be the name of the ancestor (excluding the separator).
+    substring_length = index_last_separator;
+  }
+  // Neither the logger nor its ancestors have had their severity threshold specified.
+  return g_rcutils_logging_default_severity_threshold;
 }
 
-int rcutils_logging_get_severity_threshold()
+rcutils_ret_t rcutils_logging_set_logger_severity_threshold(const char * name, int severity)
 {
-  return g_rcutils_logging_severity_threshold;
+  RCUTILS_LOGGING_AUTOINIT
+  if (!g_rcutils_logging_severities_map_valid) {
+    return RCUTILS_RET_LOGGING_SEVERITY_MAP_INVALID;
+  }
+
+  const char * severity_string;
+  if (RCUTILS_LOG_SEVERITY_DEBUG == severity) {
+    severity_string = "DEBUG";
+  } else if (RCUTILS_LOG_SEVERITY_INFO == severity) {
+    severity_string = "INFO";
+  } else if (RCUTILS_LOG_SEVERITY_WARN == severity) {
+    severity_string = "WARN";
+  } else if (RCUTILS_LOG_SEVERITY_ERROR == severity) {
+    severity_string = "ERROR";
+  } else if (RCUTILS_LOG_SEVERITY_FATAL == severity) {
+    severity_string = "FATAL";
```

Comment:
This mapping seems useful and is already defined later in this file, should we put it somewhere for reuse?

Reply:
OK, 07f733c167f6aebfb0dc9794184b6dc0542d3d6c adds a mapping and uses it for the threshold settings and 939a4352cde7d475f40a978af5476ac745f43c7f uses it for the console outputting